vmware-event-router/internal/processor/openfaas/openfaas.go

```diff
@@ -0,0 +1,322 @@
+	}
+
+	ctlCfg := ofsdk.ControllerConfig{
+		GatewayURL:               cfg.Address,
+		TopicAnnotationDelimiter: ofProcessor.topicDelimiter,
+		RebuildInterval:          ofProcessor.rebuildInterval,
+		UpstreamTimeout:          ofProcessor.gatewayTimeout,
+		AsyncFunctionInvocation:  cfg.Async,
+		PrintSync:                ofProcessor.verbose,
+	}
+
+	ctl := ofsdk.NewController(&credentials, &ctlCfg, ofProcessor.Logger)
+	ofProcessor.controller = ctl
+	ofProcessor.controller.Subscribe(&ofProcessor)
+	ofProcessor.controller.BeginMapBuilder()
+
+	// pre-populate the metrics stats
+	ofProcessor.stats = metrics.EventStats{
+		Provider:    string(config.ProcessorOpenFaaS),
+		Type:        config.EventProcessor,
+		Address:     cfg.Address,
+		Started:     time.Now().UTC(),
+		Invocations: make(map[string]*metrics.InvocationDetails),
+	}
+	go ofProcessor.PushMetrics(ctx, ms)
+
+	return &ofProcessor, nil
+}
+
+// defaultResponseHandler records metrics and handles invoker responses
+func defaultResponseHandler(of *Processor) responseFunc {
+	return func(res ofsdk.InvokerResponse) {
+		// TODO: currently we only support metrics when in sync invocation mode because
+		// we don't get a callback for async invocations
+		of.lock.Lock()
+
+		// check for existing topic entry
+		if _, ok := of.stats.Invocations[res.Topic]; !ok {
+			of.stats.Invocations[res.Topic] = &metrics.InvocationDetails{}
+		}
+
+		// record metrics
+		// note: only first invocation result is captured (no retries)
+		if isSuccessful(res.Status, res.Error) {
+			of.stats.Invocations[res.Topic].Success()
+		} else {
+			of.stats.Invocations[res.Topic].Failure()
+		}
+		of.lock.Unlock()
+
+		of.respChan <- res
+	}
+}
+
+// Process implements the stream processor interface and invokes any OpenFaaS
+// function subscribed to the passed cloud event. If the processor has already
+// been shutdown, ErrStopped will be returned.
+func (p *Processor) Process(ctx context.Context, ce cloudevents.Event) error {
+	if p.isStopped() {
+		return ErrStopped
+	}
+
+	if p.verbose {
+		p.Printf("processing event (ID %s): %v", ce.ID(), ce)
+	}
+
+	topic, message, err := handleEvent(ce)
+	if err != nil {
+		return processor.NewError(config.ProcessorOpenFaaS, errors.Wrapf(err, "handle event %v", ce))
+	}
+
+	if p.verbose {
+		p.Printf("created new outbound event for subscribers: %q", string(message))
+	}
+
+	p.Printf("invoking function(s) for event %q on topic: %q", ce.ID(), topic)
+	defer func() {
+		p.Printf("finished processing of event %q on topic: %q", ce.ID(), topic)
+	}()
+
+	m, err := p.controller.InvokeWithContext(ctx, topic, message)
+	if err != nil {
+		return processor.NewError(config.ProcessorOpenFaaS, errors.Wrap(err, "invoke function"))
+	}
+
+	p.Printf("%d function(s) matched for event %q on topic: %q", m, ce.ID(), topic)
+	if m == 0 {
+		return nil
+	}
+
+	p.Printf("waiting for %d functions to return", m)
+
+	// configure retry options
+	retryOpts := []retry.Option{
+		retry.Attempts(3),
```

Comment:
Would be nice to have these configuration values as constants.

Reply:
Done